src/doorman/doorman.cpp

```diff
@@ -0,0 +1,367 @@
+}
+
+void Doorman::poll() {
+    mud_.poll();
+    socket_poll();
+    check_for_dead_lookups();
+}
+
+void Doorman::check_for_dead_lookups() {
+    pid_t waiter;
+    int status;
+
+    do {
+        waiter = waitpid(-1, &status, WNOHANG);
+        if (waiter == -1) {
+            break; // Usually 'no child processes'
+        }
+        if (waiter) {
+            // Clear the zombie status
+            waitpid(waiter, nullptr, 0);
+            log_out("Lookup process %d died", waiter);
+            // Find the corresponding channel, if still present
+            if (auto chanIt = std::find_if(begin(channels_), end(channels_),
+                                           [&](const Channel &chan) { return chan.has_lookup_pid(waiter); });
+                chanIt != end(channels_)) {
+                chanIt->on_lookup_died(status);
+            }
+        }
+    } while (waiter);
+}
+
+void Doorman::socket_poll() {
+    fd_set input_fds, exception_fds;
+    FD_ZERO(&input_fds);
+    FD_ZERO(&exception_fds);
+    FD_SET(listenSock_.number(), &input_fds);
+    int maxFd = listenSock_.number();
+    if (mud_.fd_ok()) {
+        FD_SET(mud_.fd().number(), &input_fds);
+        FD_SET(mud_.fd().number(), &exception_fds);
+        maxFd = std::max(maxFd, mud_.fd().number());
+    }
+    for (auto &channel : channels_)
+        maxFd = std::max(channel.set_fds(input_fds, exception_fds), maxFd);
+
+    timeval timeOut = {1, 0}; // Wakes up once a second to do housekeeping
+    int nFDs = select(maxFd + 1, &input_fds, nullptr, &exception_fds, &timeOut);
+    if (nFDs == -1 && errno != EINTR) {
+        log_out("Unable to select()!");
+        perror("select");
+        exit(1);
+    }
+    /* Anything happened in the last 10 seconds? */
+    if (nFDs <= 0)
+        return;
+    /* Has the MUD crashed out? */
+    if (mud_.connected() && FD_ISSET(mud_.fd().number(), &exception_fds)) {
+        mud_.close();
+        return; /* Prevent falling into packet handling */
+    }
+    for (int i = 0; i <= maxFd; ++i) {
+        /* Kick out the freaky folks */
+        if (FD_ISSET(i, &exception_fds)) {
+            abort_connection(i);
+        } else if (FD_ISSET(i, &input_fds)) { // Pending incoming data
+            /* Is it the listening connection? */
+            if (i == listenSock_.number()) {
+                accept_new_connection();
+            } else if (mud_.fd().is_open() && i == mud_.fd().number()) {
+                mud_.process_mud_message();
+            } else {
+                /* It's a message from a user or ident - dispatch it */
+                if (auto it = std::find_if(begin(channels_), end(channels_),
+                                           [&](const Channel &chan) { return chan.is_hostname_fd(i); });
+                    it != end(channels_)) {
+                    it->incomingHostnameInfo();
+                } else {
+                    int nBytes;
+                    byte buf[256];
```

Comment:
We should perhaps have a single constant in a header for this

Reply:
Agreed; will use an appropriately-calculated amount so as not to overflow the `Channel::MaxPayload` when I put it in (noted in the PR description).